examples/training/classification/densenet/imagenet/basic_training.py

```diff
@@ -0,0 +1,224 @@
+# Copyright 2022 The KerasCV Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import sys
+
+import tensorflow as tf
+from absl import flags
+from keras.callbacks import BackupAndRestore
+from keras.callbacks import EarlyStopping
+from keras.callbacks import ModelCheckpoint
+from keras.optimizers import Adam
+from tensorflow.keras.optimizers.schedules import PolynomialDecay
+
+import keras_cv
+from keras_cv.models import DenseNet121
+
+"""
+Title: Training a DenseNet for Imagenet Classification with KerasCV
+Author: [ianjjohnson](https://github.com/ianjjohnson)
+Date created: 2022/07/25
+Last modified: 2022/07/25
+Description: Use KerasCV to train a DenseNet using modern best practices for image classification
+"""
+
+"""
+## Overview
+KerasCV makes training state-of-the-art classification models easy by providing implementations of modern models, preprocessing techniques, and layers.
+In this tutorial, we walk through training a DenseNet model against the Imagenet dataset using Keras and KerasCV.
+This tutorial requires you to have KerasCV installed:
+```shell
+pip install keras-cv
+```
+"""
+
+"""
+## Setup and flags
+
+"""
+
+flags.DEFINE_string("imagenet_path", None, "Directory from which to load Imagenet.")
+flags.DEFINE_string(
+    "backup_path", None, "Directory which will be used for training backups"
+)
+flags.DEFINE_string(
+    "weights_path", None, "Directory which will be used to store weight checkpoints."
+)
+
+FLAGS = flags.FLAGS
+FLAGS(sys.argv)
+
+NUM_CLASSES = 1000
+BATCH_SIZE = 256
+IMAGE_SIZE = (300, 300)
+EPOCHS = 250
+
+"""
+## Data loading
+This guide uses the
+[Imagenet dataset](https://www.tensorflow.org/datasets/catalog/imagenet2012). Note that this requires manual download, and does not work out-of-the-box with TFDS.
+To get started, we first load the dataset from a command-line specified directory where ImageNet is stored as TFRecords.
+"""
+
+
+def parse_imagenet_example(example, IMAGE_SIZE):
+    # Read example
+    image_key = "image/encoded"
+    label_key = "image/class/label"
+    keys_to_features = {
+        image_key: tf.io.FixedLenFeature((), tf.string, ""),
+        label_key: tf.io.FixedLenFeature([], tf.int64, -1),
+    }
+    parsed = tf.io.parse_single_example(example, keys_to_features)
+
+    # Decode and resize image
+    image_bytes = tf.reshape(parsed[image_key], shape=[])
+    image = tf.io.decode_jpeg(image_bytes, channels=3)
+    image = tf.image.resize(image, IMAGE_SIZE)
```

Comment:
keras smart_resize tends to outperform tf.image.resize in terms of final model scores if I remember correctly.

Reply:
@LukeWood If not `tf.image` maybe a `tf.keras.layers.Resizing(..., crop_to_aspect_ratio=True)` instead?

from [the docs](https://www.tensorflow.org/api_docs/python/tf/keras/preprocessing/image/smart_resize):

>Warning: [tf.keras.preprocessing.image.smart_resize](https://www.tensorflow.org/api_docs/python/tf/keras/preprocessing/image/smart_resize) is not recommended for new code. Prefer [tf.keras.layers.Resizing](https://www.tensorflow.org/api_docs/python/tf/keras/layers/Resizing), which provides the same functionality as a preprocessing layer and adds [tf.RaggedTensor](https://www.tensorflow.org/api_docs/python/tf/RaggedTensor) support. See the [preprocessing layer guide](https://www.tensorflow.org/guide/keras/preprocessing_layers) for an overview of preprocessing layers.

Reply:
ah yes, thanks Sebastian.  @ianjjohnson lets use a tf.keras.layers.Resizing(..., crop_to_aspect_ratio=True) here.

Reply:
Good call -- thank you both.